src/subnet.c

```diff
@@ -37,35 +37,86 @@
 /* lists type of subnet */
 
 splay_tree_t *subnet_tree;
+static uint32_t hash_seed;
 
 /* Subnet lookup cache */
 
-static hash_t *ipv4_cache;
-static hash_t *ipv6_cache;
-static hash_t *mac_cache;
+static uint32_t hash_function_ipv4_t(const ipv4_t *p) {
+	/*
+	This basic hash works because
+	a) Most IPv4 networks routed via tinc are not /0
+	b) Most IPv4 networks have more unique low order bits
+	*/
+	uint16_t *halfwidth = (uint16_t *)p;
+	// 10.0.x.x/16 part
+	uint32_t hash = hash_seed;
+	hash += halfwidth[1] * 0x9e370001UL;
+	// x.x.0.[0-255] part
+
+#if _____LP64_____
+	return hash ^ halfwidth[0];
+#else
+	// ensure that we have a /24 with no collisions on 32bit
+	return hash ^ ntohs(halfwidth[0]);
+#endif
+}
+
+
+static uint32_t hash_function_ipv6_t(const ipv6_t *p) {
+	uint32_t *fullwidth = (uint32_t *)p;
+	uint32_t hash = hash_seed;
+
+	for(int i = 0; i < 4; i++) {
+		hash += fullwidth[i];
+		hash *= 0x9e370001UL;
+	}
+
+	return hash;
+}
+
+static uint32_t hash_function_mac_t(const mac_t *p) {
+	uint16_t *halfwidth = (uint16_t *)p;
+	uint32_t hash = hash_seed;
+
+	for(int i = 0; i < 3; i++) {
+		hash += halfwidth[i];
+		hash *= 0x9e370001UL;
+	}
+
+	return hash;
+}
+
+hash_define(ipv4_t, SUBNET_HASH_SIZE)
+hash_define(ipv6_t, SUBNET_HASH_SIZE)
+hash_define(mac_t, SUBNET_HASH_SIZE)
+
+hash_new(ipv4_t, ipv4_cache);
+hash_new(ipv6_t, ipv6_cache);
+hash_new(mac_t, mac_cache);
+
 
-void subnet_cache_flush(void) {
-	hash_clear(ipv4_cache);
-	hash_clear(ipv6_cache);
-	hash_clear(mac_cache);
+void subnet_cache_flush_table(subnet_type_t stype) {
+	// a subnet type of mac/0 means to clear all
+	if(stype != SUBNET_IPV6) { // ipv4
+		hash_clear(ipv4_t, &ipv4_cache);
+	}
+
+	if(stype != SUBNET_IPV4) { // ipv6
+		hash_clear(ipv6_t, &ipv6_cache);
+	}
+
+	hash_clear(mac_t, &mac_cache);
```

Comment:
But this will always flush the MAC cache, even if only the IPv4 or IPv6 one needs to be flushed.

Reply:
Its how the original code works and is probably necessary. Currently there is no direct knowledge of which ip subnet_t the mac belongs to, so therefore it can't be (efficiently) cleared when the ipv4 / ipv6 config changes.

So I've kept the original behaviour here.